frmts/jp2kak/jp2kakdataset.cpp

```diff
@@ -2215,25 +2215,33 @@ JP2KAKCreateCopy( const char * pszFilename, GDALDataset *poSrcDS,
         return nullptr;
     }
 
+    const bool bReversible =
+            CSLFetchNameValue(papszOptions,"Creversible") != nullptr
+            ? CPLFetchBool( papszOptions, "Creversible", false )
+            : (dfQuality >= 99.5);
+
     std::vector<kdu_long> layer_bytes;
-    layer_bytes.resize(layer_count);
 
     const int nXSize = poSrcDS->GetRasterXSize();
     const int nYSize = poSrcDS->GetRasterYSize();
+    const double dfPixelsTotal = nXSize * static_cast<double>(nYSize);
 
-    bool bReversible = false;
-
-    if( dfQuality < 99.5 )
-    {
+    // Override Quality parameter if RATE option is used
+    // TODO: highest quality that can
+    //       be achieved with a quality factor of QUALITY and
+    //       also introducing some visual weighting to the rate-distortion
+    //       optimization objective associated with all N quality layers.
+    //       That means: combines both QUALITY and RATE
+    if(CSLFetchNameValue(papszOptions,"RATE") == nullptr) {
+        layer_bytes.resize(layer_count);
```

Comment:
there should be a if (!bReversible) test protecting lines 2236 to 2252. The current code breaks the tests in autotest/gdrivers/jp2kak.py that use QUALITY=100

Reply:
You are right. The logic is a bit different since we will execute this block of code only if the RATE parameter is not passed. 
Thus, we will evaluate via the formula line 2237 the size of each layer.
If the RATE parameter is passed, we calculate the size according to it. 
The problem is that we must not execute this block of code even if bReevrsible && RATE != nullptr. So if we want to use the old logic, we have to test the quality and the existence of the RATE parameter because it is possible to specify Creversible=no with a quality >= 99.5, and even with a RATE parameter defined.

My proposal:

```cpp
 if(CSLFetchNameValue(papszOptions,"RATE") == nullptr && !bReversible) {
```

Reply:
> if(CSLFetchNameValue(papszOptions,"RATE") == nullptr && !bReversible) {

I don't think that will work, because if using just -o QUALITY=100 for example, that would go to the else block "use RATE option". The check if (!bReversible) must be inside the then clause of " if(CSLFetchNameValue(papszOptions,"RATE") == nullptr )"

Reply:
> In that case, we should execute the second block.

ok, shouldn't the logic be ?
```
const char* pszRate = CSLFetchNameValue(papszOptions,"RATE") ;
if( pszRate != nullptr )
{
    // do what is in the current else clause
}
else if( !bReversible )
{
     // do what is in the current then clause
}
```

It would be good if you could add some testing of the the new options in autotest/gdrivers/jp2kak.py. It might be possible to the dal.GetJPEG2000Structure() to check that the effect is correct . For an example of how to use that method, see jp2openjpeg_test_codeblock() in autotest/gdrivers/jp2openjpeg.py

Reply:
Yes I think your proposal works fine. I will make the changes.
For the tests, I am not familar with pytest, I tried to run the pytest using the documentation https://github.com/OSGeo/gdal/blob/master/CONTRIBUTING.md  but many tests fail or are skipped.

Using this particular command:

```sh
pytest autotest/gdrivers/jp2kak.py
```

All tests fails because it does not find JP2KAK:

```sh
        gdaltest.jp2kak_drv = gdal.GetDriverByName('JP2KAK')
>       assert gdaltest.jp2kak_drv is not None
E       assert None is not None
E        +  where None = gdaltest.jp2kak_drv

```
Any help would be appreciated

Reply:
> All tests fails because it does not find JP2KAK:

I assume ``gdalinfo --format JP2KAK`` works ?
then try ``python -c "from osgeo import gdal; print(gdal.GetDriverByName('JP2KAK'))"``
if that returns None, then it means that either you have not built the Python bindings (cd swig/python && make) and/or your PYTHONPATH doesn't point to them . They are typically in a path like swig/python/build/lib.linux-x86_64-3.8

Reply:
Yes, that fixes the problem. Thx. I will now work on tests and push changes

Reply:
For the testing part, I tried to use what you mentioned but I cannot get the interesting part with the current implementation.
Here what I want to test `Kdu-Layer-Info: log_2{Delta-D(squared-error)/Delta-L(bytes)`:

```sh
 [0
 'Field'
 [2
 'name'
 [1
 'COM']]
 [2
 'type'
 [1
 'string']]
 [2
 'size'
 [1
 '12']]
 [1
 'Kakadu-vs8.1']]]
 [0
 'Marker'
 [2
 'name'
 [1
 'COM']]
 [2
 'offset'
 [1
 '248']]
 [2
 'length'
 [1
 '281']]
 [0
 'Field'
 [2
 'name'
 [1
 'Rcom']]
 [2
 'type'
 [1
 'uint16']]
 [2
 'description'
 [1
 'LATIN1']]
 [1
 '1']]
 [0
 'Field'
 [2
 'name'
 [1
 'COM']]
 [2
 'type'
 [1
 'string']]
 [2
 'size'
 [1
 '275']]
 [1
 'Kdu-Layer-Info: log_2{Delta-D(squared-error)/Delta-L(bytes)}
 L(bytes)\n  -6.9
  4.8e+02\n  -8.1
  5.0e+02\n  -9.3
  5.3e+02\n -10.5
  5.7e+02\n -11.6
  6.2e+02\n -12.8
  6.7e+02\n -14.0
  7.5e+02\n -15.1
  9.3e+02\n -16.3
  1.2e+03\n -17.5
  1.6e+03\n -18.7
  2.2e+03\n -21.3
  4.6e+03\n']]]
```

If I have correctly understand how it works, I have to use that structure:

```py
get_element_val(find_element_with_name(NODE1, "Field", "COM"))
```
But I am getting `Kakadu-vs8.1`, and thats normal. Either I am wrong on how to use the parser, or we have to evolve the parser to iterate on the different node instead of getting the first one.

Reply:
Fix condition logic in e475326f4201df20765df66943cc250a50ab00f3

Reply:
Add tests in e8e1dc25cf54aad9074e86cc89177e17d7052f94.
Note, the GDalStructure does not return a approximate value of the size of each layer. 
For this reason, I preferred to test the string containing these approximate values. I didn't find a way to get them otherwise because these values are passed directly to the KDU compressor

Reply:
Fix in dccef0b7037c363fe253bbdb99186f2e4b5c8e11.

Some words, I restricted the tests to test only the size of the layers calculated according to the parameter passed in RATE. Indeed, depending on the way the RATE parameter is used, the way the layers are calculated differs. For example, if we have 4 LAYERS and we pass only RATE=1.5, the first 3 layers will be 0 and the last one will be calculated by floor(dfPixelsTotal * currentBitRate * 0.125F).

The layers with a value of 0 are automatically calculated by KDU afterwards. So what is testable are the layers calculated by the formula on the GDAL driver side.

So, I restricted the tests to compare only this value. I tested with Kakadu 7.10 and Kakadu 8.1 and the tests pass. I didn't test with kakadu 6.x but I think it's better to focus on the most recent versions but it probably works very well with version 6.